subsys/net/lib/aws_jobs/src/aws_jobs.c

```diff
@@ -0,0 +1,381 @@
+ * @param[in] job_id JobId of the current accepted job.
+ *
+ */
+int subscribe_job_id_get_topic(struct mqtt_client *client, u8_t *job_id)
+{
+	struct mqtt_topic subscribe_topic = construct_job_id_get_topic(
+						client->client_id.utf8,
+						job_id
+						);
+	const struct mqtt_subscription_list subscription_list = {
+		.list = (struct mqtt_topic *)&subscribe_topic,
+		.list_count = 1,
+		.message_id = SUBSCRIBE_NOTIFY
+	};
+	LOG_DBG("Subscribe: %s\n", subscribe_topic.topic.utf8);
+
+	return mqtt_subscribe(client, &subscription_list);
+}
+
+
+/**
+ * @brief Constructs the update topic for AWS jobs.
+ *
+ * @param[in] client_id Client ID of the current connected MQTT session.
+ * @param[in] job_id The JobId of the current recived job.
+ *
+ */
+static struct mqtt_topic construct_job_id_update_topic(u8_t *client_id,
+						       u8_t *job_id)
+{
+	char job_id_update_topic[JOB_ID_UPDATE_TOPIC_MAX_LEN + 1];
+	int ret = snprintf(job_id_update_topic,
+			   JOB_ID_UPDATE_TOPIC_MAX_LEN,
+			   JOB_ID_UPDATE_TOPIC_TEMPLATE,
+			   client_id,
+			   job_id);
+
+	report_snprintf_err(ret,
+			    JOB_ID_UPDATE_TOPIC_MAX_LEN,
+			    "job_id__update_topic");
+
+	struct mqtt_topic topic = {
+		.topic = {
+			.utf8 = job_id_update_topic,
+			.size = strlen(job_id_update_topic)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+
+	return topic;
+}
+
+
+/**
+ * @brief AWS jobs update job execution status function.
+ *
+ * This implements the minimal requirements for doing an AWS job only
+ * updating status and status details update are supported.
+ *
+ * @param[in] client Initialized and connected MQTT Client instance.
+ * @param[in] job_id The ID of the job which you are updating this will be a
+ *		part of the MQTT topic used to update the job.
+ * @param[in] status The Job execution status.
+ * @param[in] status_details JSON object in stringformat containing additional
+		information. This object can contain up to 10 fields according
+		to the AWS IoT jobs documentation.
+ * @param[in] expected_version The expected job document version this needs to
+ *		be incremented by 1 for every update.
+ * @param[in] client_token This can be an arbitrary value and will be reflected
+ *		in the response to the update.
+ *
+ * @retval 0 If update is published successfuly otherwise return mqtt_publish
+ *	     error code.
+ *
+ */
+#define UPDATE_JOB_PAYLOAD "{\"status\":\"%s\",\
+			    \"statusDetails\":\"%s\",\
+			    \"expectedVersion\": \"%d\",\
+			    \"clientToken\": \"%s\"}"
+int aws_update_job_execution(struct mqtt_client *client,
+			     u8_t *job_id,
+			     enum execution_status status,
+			     u8_t *status_details,
+			     int expected_version,
+			     u8_t *client_token)
+{
+	/* Max size document is 1350 char but the max size json document is
+	 * actually 32kb set it to what is the limiting factor which is the MQTT
+	 * buffer size for reception end
+	 */
+	u8_t update_job_payload[CONFIG_MQTT_MESSAGE_BUFFER_SIZE];
```

Comment:
Bit unsure about if we should use this define or define our own. The problem is that the size limitation of the MQTT buffer limits the size of what you should be able to send as payload.

Reply:
I would say use your own buffer size but have a static assert to make sure it doesn't exceeds MQTT 's.

Reply:
Actually, you should not use `CONFIG_MQTT_MESSAGE_BUFFER_SIZE` at all here. Upstream MQTT doesn't bother with buffer size, and leaves it to the application. If you look closer, you'll see that `CONFIG_MQTT_MESSAGE_BUFFER_SIZE` is a config specific for `mqtt_simple` sample.

Reply:
Okei thanks for the input :)

Reply:
So there is no way for me to assert if the MQTT buffer is smaller than the expected payload size?

Reply:
You don't need to bother about the MQTT in this case at all. The upstream library does not copy the PUBLISH payload into internal buffers, but sends them directly from the buffer you provide. So whatever size you use here, MQTT will handle. 

The PUBLISH header though will be created in an internal MQTT buffer, which you provide during the `mqtt_client` initialization. So once you're there, you need to make sure it's large enough to fit the topic and the remaining header contents. In case the PUBLISH header does not fit into the buffer provided, `mqtt_publish` will return an error code (`-ENOMEM` afair).

Reply:
Okei, then I'll check for that in the application.